XPlat compatibility changes

ValMenn opened the pull request:
Fix slashes on Unix & Mac in properties that look like paths. If the
path points at the Mono framework directory, fix it up based on the
framework version.

Add some of the standard properties associated with toolsets. These
are normally found in the registry, but for Mono they are easy to
deduce.

dnfclas commented:
Hi __@ValMenn__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

sarajoiner commented on src/XMakeBuildEngine/Instance/ProjectInstance.cs:
```diff
@@ -2242,7 +2242,7 @@ private void Initialize(ProjectRootElement xml, IDictionary<string, string> glob
             if (NativeMethodsShared.IsUnixLike && !_globalProperties.Contains("MSBuildFrameworkToolsRoot"))
             {
-                _globalProperties.Set(ProjectPropertyInstance.Create("MSBuildFrameworkToolsRoot", NativeMethodsShared.FrameworkBasePath, false));
+                _globalProperties.Set(ProjectPropertyInstance.Create("MSBuildFrameworkToolsRoot", NativeMethodsShared.FrameworkBasePath + '/', false));
```
Seems like now that you're doing all that extra work in ToolsetReader to set all the other properties, that would be a better place for this than shoving it into the global properties here.

sarajoiner commented on src/XMakeBuildEngine/Definition/ToolsetReader.cs:
```diff
@@ -471,6 +486,113 @@ bool accumulateProperties
         }
 
         /// <summary>
+        /// Create a dictionary with standard properties.
```
Does Mono support System.Configuration?  Seems like it would be a lot more data-driven (and less code) to just add toolsets via the .exe.config rather than having to hard-code all of this if so.

sarajoiner commented on src/XMakeBuildEngine/Definition/ToolsetReader.cs:
```diff
@@ -471,6 +486,113 @@ bool accumulateProperties
+                    properties.Set(
+                        ProjectPropertyInstance.Create("MSBuildFrameworkToolsPath32", toolsPath, true, false));
+                }
+                /*if (globalProperties["MSBuildToolsPath"] == null && properties["MSBuildToolsPath"] == null)
```
nit: please don't check in commented out code (at least not without a description of why)

ValMenn commented:
The toolset properties are specified in the registry on Windows. They are toolset-specific. Placing them into the .exe.config has problems. These are identified below.
1. They are computed during runtime. I am not sure I can express them in terms of other properties so they can be evaluated.
2. They are Mono-specific. I did not want to place platform-specific things into a place where I cannot identify them as such.

I could create a mono-specific config file, but see issue 1 above. Plus, that would not be less code and I would have to either express them in toolset-independent way or have the same properties for all toosets, which would make expressing them in terms of other properties even more complicated.

Finally, I commented out two of the properties, because they are reserved and I could not add them to the dictionary. I am still to find a good place to generate them, possibly along with the MSBuildFrameworkToolsRoot, which is not a part of a toolset definition and was not, therefore, added to the toolset-specific set.

sarajoiner commented:
In addition to being able to depend on environment properties, global properties, or previously defined toolset properties, I believe toolset properties can also be defined in terms of property functions, so you should be able to e.g. define a toolset property in terms of, say, the output of one of the ToolLocationHelper property functions.  

My primary concern with hard-coding everything inside MSBuild is that if any user wished to modify the toolset, now they have no ability to do so without recompiling MSBuild.  This is very common for checked-in build systems.  

... On the other hand, recompiling MSBuild is a much lower bar these days than it used to be. So maybe it doesn't matter as much anymore. 

MSBuildToolsPath is the primary toolset property, without which you would not have a toolset; as such it's passed separately, as a parameter to the Toolset() constructor, not added to the property dictionary.  

Not sure what you mean by "MSBuildFrameworkToolsRoot" not being part of a toolset definition.  Aside from MSBuildToolsPath, a toolset definition is just a bag of properties.  It can contain any property you want to throw in the bag.  (Except, of course, that it has to be something that's definable only in terms of the much more limited set of things that toolset properties can depend on.)  If you want it to be part of the toolset, go ahead and add it. :)

ValMenn commented:
I am still not 100% sure what the problem is. You can still define a custom toolset with all the properties in your config file as before. The defaults will be applied for global values (such as where standard toolsets are installed), but the values from the config file will not be overwritten. In addition, standard properties (which are discoverable on Mono without a registry) will be applied to standard toolsets (just like under Windows, except not from the registry). I think this behavior is as close to the Windows behavior as possible, considering that the data from the registry is not available).

JeremyKuhne commented on src/Shared/FileUtilities.cs:
```diff
@@ -421,6 +421,72 @@ internal static string FixFilePath(string path)
+        /// Use a native shared method to massage file path. If the file is adjusted,
+        /// that qualifies is as a path.
+        /// </summary>
+        internal static string MaybeAdjustFilePath(string value)
```
Would AdjustForFileSystem() be clearer?

ValMenn replied:
Possibly, but I wanted to emphasize that it's not just for the FileSystem and it does not process every string, only whatever looks like path (Maybe seems like a good indicator, but it may just be due to my lack of imagination). In addition to the file system related adjustment (slashes) it may also correct a path for a framework file on Mono.

JeremyKuhne commented on src/Shared/FileUtilities.cs:
```diff
@@ -421,6 +421,72 @@ internal static string FixFilePath(string path)
+            }
+
+            // For Unix-like systems, we may want to convert backslashes to slashes
+            string newValue = Regex.Replace(value, @"[\\/]+", "/");
```
StringBuilder would likely be much more performant through this method. Faster than Regex and less string allocations.

ValMenn replied:
First, I think regex would only create one (output) string. Perf did not seem to be an issue. I ran a 1 min build with and without the function (for the run without the function I hand-fixed projects to have valid paths). The time was pretty much the same (few seconds difference was likely a precision error, especially considering that the build that actually changed the paths was a few sec faster).

JeremyKuhne replied:
Using StringBuilder you could avoid at least a couple string allocations when you go through, particularly if you pass it to the other fixup method and used a cached one. I don't care too much if we don't expect a lot of properties with any forward or back slashes. (Incidentally, why are you replacing forward slash?) By far the biggest impact in this method is the Directory.Exists I'm sure.

sarajoiner replied:
FWIW, time spent inside the MSBuild engine is usually a pretty small proportion of the overall build time, so on a one-minute build I would expect even a real regression not to show up as any sort of human-visible difference.  (Unless the regression was excessively large, of course.  Or unless you put together a fake project that did nothing but exercise the changed codepath.) 

Whether or not it's worth digging deeper into whether / how much this codepath has regressed probably depends on what the performance bar is.

JeremyKuhne replied:
@sarajoiner Good to know. How important is memory pressure in the engine?

sarajoiner replied:
@JeremyKuhne It depends. :)  If you're building a small set of projects, I doubt anyone would notice or care except the people tracking performance.  If you're trying to build tens of thousands of projects, it matters a LOT -- one of the biggest reasons MSBuild has such scalability issues is that its dynamic work discovery makes it so that it "can't" (= can't without significant re-architecting) throw away any project configuration information or target outputs, because it doesn't know what might want to consume them later in the mind.  

It's got a pretty simple cache to disk of those two sets of things, so it'll just slow way way way down paging rather than crash with OOM, but once you hit that point, it spends so much time writing to disk, GC'ing, and reading from disk, that the build becomes effectively unusable.  And the more memory pressure there is, the faster it hits that point.  

(That said, in this case since the strings ought to just be left floating to be GC'd, I doubt it would contribute much if at all to overall memory consumption; seems more likely that there'd be a small degradation in speed due to extra time spent in Regex + extra time spent GC'ing.)

JeremyKuhne commented on src/Shared/FileUtilities.cs:
```diff
@@ -421,6 +421,72 @@ internal static string FixFilePath(string path)
+            if (NativeMethodsShared.IsWindows || string.IsNullOrWhiteSpace(value) ||
+                value.StartsWith("$(") || value.StartsWith("@(") || value.StartsWith("\\\\"))
+            {
+                return value;
```
Can you bail if there are no slashes?

ValMenn replied:
Since I change more than just fix slashes, I could only bypass the regex when there are no backslashes -- I would then have to check if the result has no forward slashes to skip the rest. Most strings that pass the first few tests are in fact paths, so that does not seem to gain much.

JeremyKuhne replied:
I don't know what one can expect to see as far as number of properties going through here. I assume 1000 isn't abnormal? 10,000?

sarajoiner commented on src/XMakeBuildEngine/Definition/ToolsetReader.cs:
```diff
@@ -471,6 +486,113 @@ bool accumulateProperties
+                {
+                    properties.Set(ProjectPropertyInstance.Create("MSBuildToolsRoot", root, true, false));
+                }
+                if (globalProperties["FrameworkSDKRoot"] == null && properties["FrameworkSDKRoot"] == null)
```
Isn't this a duplicate of above?

sarajoiner commented on src/XMakeBuildEngine/Definition/ToolsetReader.cs:
```diff
@@ -471,6 +486,113 @@ bool accumulateProperties
+                {
+                    properties.Set(ProjectPropertyInstance.Create("MSBuildToolsPath32", toolsPath, true, false));
+                }*/
+                if (globalProperties["MSBuildRuntimeVersion"] == null && properties["MSBuildRuntimeVersion"] == null)
```
FWIW I believe MSBuildRuntimeVersion is 4.0.30319 for all modern (4.0+) toolsets.

sarajoiner commented:
Re: hard-coding, your explanations make sense.  I think I was misreading the context before.  (I'm still not a fan of baking that much information into the code directly, but only on the level of a philosophical argument.)

sarajoiner commented on src/XMakeBuildEngine/Evaluation/Expander.cs:
```diff
@@ -253,6 +253,7 @@ internal string ExpandIntoStringLeaveEscaped(string expression, ExpanderOptions
             result = PropertyExpander<P>.ExpandPropertiesLeaveEscaped(result, _properties, options, elementLocation, _usedUninitializedProperties);
             result = ItemExpander.ExpandItemVectorsIntoString<I>(this, result, _items, options, elementLocation);
+            result = FileUtilities.MaybeAdjustFilePath(result);
 
```
Does this work if you have e.g. an expression that expands to something like "a\b\c.txt;d\e\f.txt" ?  Perhaps add some unit tests of the function itself?

ValMenn commented:
Addressed some of the comments. Removed commented lines, moved setting of frameworkToolsRoot to a better place in Evaluator.cs, where other built-in properties are set. Checked for slashes in MaybeAdjustFilePath.

mmitche commented:
test this please
